src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+	if (value & LONGTYPE(0xffff00000000))
+	{
+		temp = (UINT32)(value >> 16);
+		exp -= 16;
+	}
+	else
+		temp = (UINT32)value;
+
+	/* if the resulting value is 0, the reciprocal is infinite */
+	if (GCC_UNLIKELY(temp == 0))
+	{
+		*log2 = 1000 << LOG_OUTPUT_PREC;
+		return neg ? 0x80000000 : 0x7fffffff;
+	}
+
+	/* determine how many leading zeros in the value and shift it up high */
+	lz = count_leading_zeros(temp);
+	temp <<= lz;
+	exp += lz;
+
+	/* compute a pointer to the table entries we want */
+	/* math is a bit funny here because we shift one less than we need to in order */
+	/* to account for the fact that there are two UINT32's per table entry */
+	table = &voodoo_reciplog[(temp >> (31 - RECIPLOG_LOOKUP_BITS - 1)) & ((2 << RECIPLOG_LOOKUP_BITS) - 2)];
+
+	/* compute the interpolation value */
+	interp = (temp >> (31 - RECIPLOG_LOOKUP_BITS - 8)) & 0xff;
+
+	/* do a linear interpolatation between the two nearest table values */
+	/* for both the log and the reciprocal */
+	rlog = (table[1] * (0x100 - interp) + table[3] * interp) >> 8;
+	recip = (table[0] * (0x100 - interp) + table[2] * interp) >> 8;
+
+	/* the log result is the fractional part of the log; round it to the output precision */
+	rlog = (rlog + (1 << (RECIPLOG_LOOKUP_PREC - LOG_OUTPUT_PREC - 1))) >> (RECIPLOG_LOOKUP_PREC - LOG_OUTPUT_PREC);
+
+	/* the exponent is the non-fractional part of the log; normally, we would subtract it from rlog */
+	/* but since we want the log(1/value) = -log(value), we subtract rlog from the exponent */
+	*log2 = ((exp - (31 - RECIPLOG_INPUT_PREC)) << LOG_OUTPUT_PREC) - rlog;
+
+	/* adjust the exponent to account for all the reciprocal-related parameters to arrive at a final shift amount */
+	exp += (RECIP_OUTPUT_PREC - RECIPLOG_LOOKUP_PREC) - (31 - RECIPLOG_INPUT_PREC);
+
+	/* shift by the exponent */
+	if (exp < 0)
+		recip >>= -exp;
+	else
+		recip <<= exp;
+
+	/* on the way out, apply the original sign to the reciprocal */
+	return neg ? -(INT64)recip : (INT64)recip;
+}
+
+
+
+/*************************************
+ *
+ *  Float-to-int conversions
+ *
+ *************************************/
+
+inline INT32 float_to_int32(UINT32 data, int fixedbits)
+{
+	int exponent = ((data >> 23) & 0xff) - 127 - 23 + fixedbits;
+	INT32 result = (data & 0x7fffff) | 0x800000;
+	if (exponent < 0)
+	{
+		if (exponent > -32)
+			result >>= -exponent;
+		else
+			result = 0;
+	}
+	else
+	{
+		if (exponent < 32)
+			result <<= exponent;
+		else
+			result = 0x7fffffff;
+	}
+	if (data & 0x80000000)
+		result = -result;
+	return result;
+}
+
+inline INT64 float_to_int64(UINT32 data, int fixedbits)
+{
+	int exponent = ((data >> 23) & 0xff) - 127 - 23 + fixedbits;
+	INT64 result = (data & 0x7fffff) | 0x800000;
+	if (exponent < 0)
+	{
+		if (exponent > -64)
+			result >>= -exponent;
+		else
+			result = 0;
+	}
+	else
+	{
+		if (exponent < 64)
+			result <<= exponent;
+		else
+			result = LONGTYPE(0x7fffffffffffffff);
+	}
+	if (data & 0x80000000)
+		result = -result;
```

Comment:
Same problems here as above in the 32-bit case. I hope we avoid using these and let the compiler do a fast and correct (saturated) conversion to int64.

Reply:
Acknowledged.